ragna/embedding_models/_all_minilm_l6_v2.py

```diff
@@ -0,0 +1,103 @@
+from collections import deque
+from typing import Deque, Iterable, Iterator, TypeVar, Union
+from uuid import UUID
+
+from ragna.core import (
+    Chunk,
+    Document,
+    Embedding,
+    EmbeddingModel,
+    PackageRequirement,
+    Page,
+    Requirement,
+)
+
+T = TypeVar("T")
+
+
+# The function is adapted from more_itertools.windowed to allow a ragged last window
+# https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.windowed
+def windowed_ragged(
+    iterable: Iterable[T], *, n: int, step: int
+) -> Iterator[tuple[T, ...]]:
+    window: Deque[T] = deque(maxlen=n)
+    i = n
+    for _ in map(window.append, iterable):
+        i -= 1
+        if not i:
+            i = step
+            yield tuple(window)
+
+    if len(window) < n:
+        yield tuple(window)
+    elif 0 < i < min(step, n):
+        yield tuple(window)[i:]
+
+
+class AllMiniLML6v2(EmbeddingModel):
+    @classmethod
+    def display_name(cls) -> str:
+        return "all-MiniLM-L6-v2"
+
+    @classmethod
+    def requirements(cls) -> list[Requirement]:
+        return [
+            # Rather than using sentence transformers package by itself, chromadb
+            # packaged this embedding model and is a much lighter dependency.
+            PackageRequirement("chromadb>=0.4.13"),
+            PackageRequirement("tiktoken"),
+        ]
+
+    def __init__(self) -> None:
+        super().__init__()
+        import tiktoken
+        from chromadb.utils import embedding_functions
+
+        self._tokenizer = tiktoken.get_encoding("cl100k_base")
+        self._model = embedding_functions.ONNXMiniLM_L6_V2()
+
+    def embed_documents(self, documents: list[Document]) -> list[Embedding]:
+        return [
+            Embedding(embedding=self.embed_text(chunk.text), chunk=chunk)
+            for document in documents
+            for chunk in self._chunk_pages(
+                document.extract_pages(),
+                document_id=document.id,
+                chunk_size=500,
+                chunk_overlap=250,
+            )
+        ]
+
+    def embed_text(
```

Comment:
After looking at the PR in more detail, I again have my doubts about this function. As discussed before, this only exists for our benefit of embedding the prompt. However it comes at the cost of

1. We need to mess with the type annotations quite a bit, because the typing system doesn't really support output types based on the input, e.g. `str -> str` and `list[str] -> list[str]`.
2. Users have to implement two methods instead of one.
3. We need to integrate two very similar methods into the Ragna protocol.

I think the costs outweigh the benefit here.

Right now `embed_documents` takes `Document`s. But this will change with the chunking work to take `Chunk`s. If we move the chunking into `Chat.prepare` for now to emulate the future case for the embedding model, couldn't we create a private `_embed_text` method stuffs the text into a dummy `Chunk`, calls the user defined `embed_documents` and returns the embedding? That way we should be able to eliminate all the downsides above while keeping the benefit.

Thoughts?

Reply:
I think, after reading a good analysis you are correct.  I will implement your solution.

Reply:
The main thing swaying it for me is that the user (ragna developer) has to implement two different methods.  I think this warrants the roundabout method.

Reply:
I'd also like to point out that pretty much no matter what you do, you'll still encounter the issue of passing through a batch vs passing through a single object.  We really want to be able to batch chunks because it is simply more efficient to batch-encode these things.  I will show you the difference after finishing up these comments.

Reply:
You can take a look at what I've done in commit [8c9fb84](https://github.com/Quansight/ragna/pull/369/commits/8c9fb8436f79e8622eac51c06ad2a3003f4644a3), I don't particularly like it but it does work.